File: downloader/core/src/commonMain/kotlin/com/flowspeed/lib/downloader/downloaditem/http/HttpDownloader.kt
package com.flowspeed.lib.downloader.downloaditem.http

import com.flowspeed.lib.downloader.DownloadManager
import com.flowspeed.lib.downloader.Downloader
import com.flowspeed.lib.downloader.connection.HttpDownloaderClient
import com.flowspeed.lib.downloader.downloaditem.IDownloadItem
import kotlinx.serialization.KSerializer
import kotlin.reflect.KClass

class HttpDownloader(
    httpDownloaderClient: Lazy<HttpDownloaderClient>
) : Downloader<HttpDownloadItem, HttpDownloadJob, HttpDownloadCredentials> {
    val httpDownloaderClient by httpDownloaderClient
    override fun createJob(
        item: HttpDownloadItem,
        downloadManager: DownloadManager,
    ): HttpDownloadJob {
        return HttpDownloadJob(
            item,
            downloadManager,
            httpDownloaderClient,
        )
    }

    override fun accept(item: IDownloadItem): Boolean {
        return item is HttpDownloadItem
    }

    override val downloadItemClass: KClass<HttpDownloadItem> = HttpDownloadItem::class
    override val downloadCredentialsClass: KClass<HttpDownloadCredentials> = HttpDownloadCredentials::class
    override val downloadJobClass: KClass<HttpDownloadJob> = HttpDownloadJob::class
    override val downloadItemSerializer: KSerializer<HttpDownloadItem> = HttpDownloadItem.serializer()
    override val downloadCredentialsSerializer: KSerializer<HttpDownloadCredentials> =
        HttpDownloadCredentials.serializer()
}

File: FlowDownloadManager/src/Flow.Core/Models/HttpDownloader.cs
using System;

namespace Flow.Core.Models;

public class HttpDownloader : IDownloader<HttpDownloadItem, HttpDownloadJob, HttpDownloadCredentials>
{
    private readonly Lazy<Connection.HttpDownloaderClient> _clientLazy;

    public Connection.HttpDownloaderClient Client => _clientLazy.Value;

    public HttpDownloader(Lazy<Connection.HttpDownloaderClient> clientLazy)
    {
        _clientLazy = clientLazy ?? throw new ArgumentNullException(nameof(clientLazy));
    }

    public DownloadJob CreateJob(IDownloadItem item, DownloadManager downloadManager)
    {
        if (item is not HttpDownloadItem httpItem)
        {
            throw new ArgumentException($"Expected HttpDownloadItem but got {item?.GetType().FullName}", nameof(item));
        }
        return new HttpDownloadJob(httpItem, downloadManager, Client);
    }

    public bool Accept(IDownloadItem item)
    {
        return item is HttpDownloadItem;
    }

    public Type DownloadItemType => typeof(HttpDownloadItem);
    public Type DownloadCredentialsType => typeof(HttpDownloadCredentials);
    public Type DownloadJobType => typeof(HttpDownloadJob);
}
